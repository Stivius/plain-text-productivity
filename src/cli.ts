#! /usr/bin/env node

import main from "./index";

main().catch((err) => console.log(err, "Error while running"));
