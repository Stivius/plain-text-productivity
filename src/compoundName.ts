const SEPARATOR = "/";

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundName {
  parent: CompoundName;
  name: string;

  constructor(value: string) {
    const splitted = value.split(SEPARATOR);
    if (splitted.length > 0) {
      const name = splitted.pop();
      if (name.length === 0) {
        throw Error("subpart cannot be empty");
      }
      if (splitted.length > 0) {
        this.parent = new CompoundName(splitted.join(SEPARATOR));
      }
      this.name = name;
    } else {
      this.name = value;
    }
  }

  toString(): string {
    if (this.parent != null) {
      return `${this.parent.toString()}/${this.name}`;
    }
    return this.name;
  }

  isEqual(other: CompoundName): boolean {
    return this.toString() === other.toString();
  }
}

export const searchForCompoundProject = (searchedValue: CompoundName) => {
  return (value: CompoundName) => value.isEqual(searchedValue);
};
