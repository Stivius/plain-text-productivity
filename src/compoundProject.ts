const SEPARATOR = '/'

// if has children than can't be used as record
// EstimatedProject contain only leafs

export class CompoundProject {
    parent: CompoundProject;
    name: string;

    constructor(value: string) {
        const splitted = value.split(SEPARATOR);
        if (splitted.length > 0) {
            const name = splitted.pop();
            if (name.length === 0) {
                throw Error('subpart cannot be empty');
            }
            if (splitted.length > 0) {
                this.parent = new CompoundProject(splitted.join(SEPARATOR));
            }
            this.name = name;
        } else {
            this.name = value;
        }
    }

    toString(): string {
        if (this.parent != null) {
            return `${this.parent.toString()}/${this.name}`
        }
        return this.name;
    }

    isEqual(other: CompoundProject): boolean {
        return this.toString() === other.toString();
    }
}
