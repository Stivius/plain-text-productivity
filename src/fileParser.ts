import fs from 'fs';
import moment from 'moment';
import {
    FileData,
    ProjectRecord,
    Mark,
    EstimatedProject,
    DATE_REGEX,
    PROJECT_NAME_REGEX,
    METADATA_PROJECT_NAME_REGEX,
    MARK_REGEX
} from "./interfaces";

enum ParsingState {
    Init = 'init',
    MetadataStarted = 'metadata_started',
    MetadataFinished = 'metadata_finished',
    MetadataActiveProjectsStarted = 'metadata_active_projects_started',
    MetadataActiveProject = 'metadata_active_projects',
    MetadataArchivedProjectsStarted = 'metadata_archived_projects_started',
    MetadataArchivedProject = 'metadata_archived_projects',
    Record = 'record'
}

type StateAction = (line: string) => void;
type StateTransitionCondition = (line: string) => boolean;

interface Transition {
    newState: ParsingState;
    condition: StateTransitionCondition;
}

class ParsingStateMachine {
    private state = ParsingState.Init;
    private stateActions = new Map<ParsingState, StateAction>();
    private stateTransitions = new Map<ParsingState, Transition[]>();

    addStateAction(state: ParsingState, action: StateAction) {
        this.stateActions.set(state, action);
    }

    addStateTransition(state: ParsingState, newState: ParsingState, condition: StateTransitionCondition) {
        if (!this.stateTransitions.has(state)) {
            this.stateTransitions.set(state, []);
        }
        let transitions = this.stateTransitions.get(state);
        transitions.push({ 
            newState: newState,
            condition: condition
        });
    }

    parseLine(line: string) {
        const transitions = this.stateTransitions.get(this.state);
        if (transitions !== undefined) {
            for (const transition of transitions) {
                if (transition.condition(line)) {
                    this.state = transition.newState;  
                    break;
                }
            }
        }
        this.performActionForState(line);
    }
    
    private performActionForState(line: string) {
        const action = this.stateActions.get(this.state);
        if (action !== undefined) {
            action(line);
        }
    }
}

export class FileParser {
    private activeProjects: string[] = [];
    private acrhivedProjects: string[] = [];
    private records: ProjectRecord[] = [];
    private currentRecord: ProjectRecord = undefined;

    constructor(private filename: string) { }

    parse(): FileData {
        this.cleanup();

        const data = fs.readFileSync(this.filename);
        const lines = data.toString().split('\n');
        const stateMachine = new ParsingStateMachine();

        stateMachine.addStateAction(
            ParsingState.Record,
            (line: string) => {
                if (this.isStartOfNewRecord(line)) {
                    if (this.currentRecord !== undefined) {
                        this.records.push(this.currentRecord);
                    }
                    this.currentRecord = this.createRecord(line);
                } else {
                    this.currentRecord.projects.push(this.parseProject(line));
                }
            }
        );

        stateMachine.addStateAction(
            ParsingState.MetadataActiveProject,
            (line: string) => {
                this.activeProjects.push(this.parseMetadataProject(line));
            }
        );

        stateMachine.addStateAction(
            ParsingState.MetadataArchivedProject,
            (line: string) => {
                this.acrhivedProjects.push(this.parseMetadataProject(line));
            }
        );

        stateMachine.addStateTransition(
            ParsingState.Init,
            ParsingState.MetadataStarted,
            (line: string) => {
                return this.isMetadataSeparator(line);
            }
        );

        stateMachine.addStateTransition(
            ParsingState.Init,
            ParsingState.Record,
            (line: string) => {
                return this.isStartOfNewRecord(line);
            }
        );

        stateMachine.addStateTransition(
            ParsingState.MetadataStarted,
            ParsingState.MetadataActiveProjectsStarted,
            (line: string) => {
                return this.isStartOfActiveProjects(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataStarted,
            ParsingState.MetadataArchivedProjectsStarted,
            (line: string) => {
                return this.isStartOfArchivedProjects(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataArchivedProject,
            ParsingState.MetadataActiveProjectsStarted,
            (line: string) => {
                return this.isStartOfActiveProjects(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataActiveProject,
            ParsingState.MetadataArchivedProjectsStarted,
            (line: string) => {
                return this.isStartOfArchivedProjects(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataActiveProjectsStarted,
            ParsingState.MetadataActiveProject,
            () => {
                return true;
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataArchivedProjectsStarted,
            ParsingState.MetadataArchivedProject,
            () => {
                return true;
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataArchivedProject,
            ParsingState.MetadataFinished,
            (line: string) => {
                return this.isMetadataSeparator(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataActiveProject,
            ParsingState.MetadataFinished,
            (line: string) => {
                return this.isMetadataSeparator(line);
            }
        )

        stateMachine.addStateTransition(
            ParsingState.MetadataFinished,
            ParsingState.Record,
            (line: string) => {
                return this.isStartOfNewRecord(line);
            }
        )

        for (const line of lines) {
            const trimmedLine = line.trim();
            if (trimmedLine.length === 0)
                continue;

            stateMachine.parseLine(trimmedLine);
        }
        if (this.currentRecord !== undefined) {
            this.records.push(this.currentRecord);
        }
        return { 
            metadata: { 
                activeProjects: this.activeProjects,
                acrhivedProjects: this.acrhivedProjects 
            }, 
            records: this.records 
        };
    }

    private cleanup() {
        this.acrhivedProjects = [];
        this.acrhivedProjects = [];
        this.records = [];
        this.currentRecord = undefined;
    }

    private isMetadataSeparator(line: string) {
        return line === '---';
    }

    private isStartOfNewRecord(line: string) {
        return line.match(DATE_REGEX) !== null;
    }

    private isStartOfActiveProjects(line: string) {
        return line.indexOf('Projects:') !== -1;
    }

    private isStartOfArchivedProjects(line: string) {
        return line.indexOf('Archived:') !== -1;
    }

    private createRecord(line: string): ProjectRecord {
        if (!moment(line, 'YYYY-MM-DD').isValid()) {
            throw Error('invalid date')
        }
        return {
            day: new Date(line),
            projects: [],
        }
    }

    private parseMetadataProject(line: string): string {
        if (!line.match(METADATA_PROJECT_NAME_REGEX)) {
            throw Error('invalid metadata project');
        }
        const parsedProject = line.replace('-', '').trim();
        if (this.activeProjects.indexOf(parsedProject) !== -1 || 
            this.acrhivedProjects.indexOf(parsedProject) !== -1) {
            throw Error('duplicated metadata project');
        }
        return parsedProject;
    }

    private parseProject(line: string): EstimatedProject {
        const splitted = line.split(':');
        const [ name, mark ] = splitted;
        if (splitted.length !== 2 || this.currentRecord === undefined || !name.match(PROJECT_NAME_REGEX)) {
            throw new Error('invalid record');
        }
        if (this.currentRecord.projects.find((value) => value.name === name) !== undefined) {
            throw new Error('duplicated project record');
        }
        if (this.activeProjects.indexOf(name) === -1 && this.acrhivedProjects.indexOf(name) === -1) {
            throw new Error('unknown project');
        }
        return {
            name: splitted[0],
            mark: this.parseMark(mark),
        };
    }

    private parseMark(value: string): Mark {
        const trimmedValue = value.trim();
        if (trimmedValue.match(MARK_REGEX) === null) {
            throw new Error('invalid mark')
        }
        return trimmedValue === '-' ? 
            undefined : 
            parseInt(trimmedValue) as Mark;
    }
}
