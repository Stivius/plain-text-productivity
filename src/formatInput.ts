import chalk from "chalk";
import columnify from "columnify";
import dateFormat from "dateformat";
import prompt from "prompt";

function formatDate(date: Date) {
  return dateFormat(date, "yyyy-mm-dd");
}

export function formatInputForFile(date: Date, data: prompt.Properties) {
  return `${formatDate(date)}\n`
    .concat(
      Object.keys(data)
        .map((k) => {
          return `${k}:${data[k]}`;
        })
        .join("\n")
    )
    .concat("\n\n");
}

export function formatInputForConsole(
  date: Date,
  data: prompt.Properties
): string {
  const colorizeMark = (value: string) => {
    switch (value) {
      case "1":
        return chalk.rgb(150, 0, 0).bold(value);
      case "2":
        return chalk.rgb(255, 125, 125).bold(value);
      case "3":
        return chalk.rgb(255, 255, 75).bold(value);
      case "4":
        return chalk.rgb(125, 255, 125).bold(value);
      case "5":
        return chalk.rgb(0, 180, 0).bold(value);
    }
    return chalk.white(value);
  };

  return `${chalk.blue(formatDate(date))}\n`.concat(
    columnify(
      Object.keys(data).map((k) => ({
        project: chalk.yellow(k),
        mark: colorizeMark(data[k] as string),
      }))
    )
  );
}
