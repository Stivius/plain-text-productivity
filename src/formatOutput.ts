import chalk from 'chalk';
import columnify from 'columnify'

export function formatDataForFile(data) {
	return Object.keys(data).map((k) => {
		if (k !== 'Date') {
			return `${k}:${data[k]}`
		} else {
			return `${data[k]}`;
		}
	}).join('\n').concat('\n');
}

export function outputDataToConsole(data) {
	const colorizeMark = (value: string) => {
		switch (value) {
			case '1': return chalk.rgb(150, 0, 0).bold(value); 
			case '2': return chalk.rgb(255, 125, 125).bold(value); 
			case '3': return chalk.rgb(255, 255, 75).bold(value); 
			case '4': return chalk.rgb(125, 255, 125).bold(value); 
			case '5': return chalk.rgb(0, 180, 0,).bold(value); 
		}
		return chalk.white(value);
	};

	const DATE_KEY = 'Date';
	if (data.hasOwnProperty(DATE_KEY)) {
		console.log(`${chalk.blue(data[DATE_KEY])}`);
		delete data[DATE_KEY];
	}

	console.log(columnify(Object.keys(data).map((k) => ({
		project: chalk.yellow(k),
		mark: colorizeMark(data[k] as string),
	}))));
}

