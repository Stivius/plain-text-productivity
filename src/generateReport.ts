import chain, {filter, round, sumBy} from 'lodash'
import {FileData, ProductivityReport, ProductivityReportOptions} from './interfaces'

export function generateReport(data: FileData, options: ProductivityReportOptions): ProductivityReport {
    const groupedData = chain(data.records)
    .filter((r) => r.day >= options.from && r.day <= options.to)
    .map((r) => r.projects)
    .flatMap()
    .filter((p) => data.metadata.activeProjects.indexOf(p.name) !== -1)
    .groupBy((assesed) => assesed.name);
    const averageMarkByProject = groupedData.map((group, groupName) => {
        const groupSum = sumBy(group, (assesed) => assesed.mark);
        const filteredGroup = filter(group, (assesed) => assesed.mark !== undefined);
        return { name: groupName, mark: groupSum / filteredGroup.length };
    });
    const maxMarkFromAverages = averageMarkByProject.map(p => p.mark - 1).max();
    const items = averageMarkByProject.map((value) => {
        const mark = value.mark - 1;
        const MAX_MARK = 4;
        if (options.absolute) {
            return { name: value.name, productivity: round(mark / MAX_MARK, 4) };
        }
        return { name: value.name, productivity: round(mark / maxMarkFromAverages, 4) };
    }).value();
    return {
        items: items,
        options: options
    };
}
