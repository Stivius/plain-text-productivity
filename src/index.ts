import commandLineArgs from 'command-line-args';
import fs from 'fs';
import {readDateFromConsole, readRecordFromConsole} from './readRecordFromInput';
import chalk from 'chalk';
import {formatInputForFile, formatInputForConsole} from './formatInput';
import {FileData, FILE_NAME} from './interfaces';
import {readRangeChoiceFromConsole} from './readRangeChoiceFromInput';
import {getRangeByType} from './getRangeByType';
import {FileParser} from './fileParser';
import {generateReport} from './generateReport';
import {formatReport} from './formatReport';

const commandsDefinitions = [
    { name: 'command', defaultOption: true },
]

const commands = commandLineArgs(commandsDefinitions, { stopAtFirstUnknown: true });

async function addRecord(date: Date, data: FileData) {
    const enteredData = await readRecordFromConsole(data.metadata.activeProjects);
    fs.appendFileSync(FILE_NAME, formatInputForFile(date, enteredData));
    console.log(formatInputForConsole(date, enteredData));
}

async function main() {
    try {
        const parser = new FileParser(FILE_NAME);
        const data: FileData = parser.parse();

        switch (commands.command) {
            case 'add':
                await addRecord(new Date(), data);
                break;

            case 'addp':
                const enteredDate = await readDateFromConsole();
                await addRecord(enteredDate, data);
                break;

            case 'lsp':
                console.log('All projects:');
                console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
                console.log(chalk.gray(data.metadata.acrhivedProjects.join('\n')));
                break;

            case 'lspa':
                console.log('Active projects:');
                console.log(chalk.yellow(data.metadata.activeProjects.join('\n')));
                break;

            case 'lspar':
                console.log('Archived projects:');
                console.log(chalk.yellow(data.metadata.acrhivedProjects.join('\n')));
                break;

            case 'report': {
                const optionDefinitions = [
                    { name: 'absolute', type: Boolean, defaultValue: false },
                    { name: 'range-type', type: Number },
                    { name: 'from-date', type: String },
                    { name: 'to-date', type: String },
                ]
                const options = commandLineArgs(optionDefinitions, { argv: commands._unknown || [] });
                const getRange = (options: commandLineArgs.CommandLineOptions) => {
                    const rangeType = options['range-type'];
                    if (rangeType === undefined) {
                        return readRangeChoiceFromConsole();
                    } else {
                        // TODO: can be entered as string and converted to number
                        if (rangeType === 0) {
                            const from = new Date(options['from-date']);
                            const to = new Date(options['to-date']);
                            if (!from || !to) {
                                throw new Error('Date range is not specified or invalid');
                            }
                            return { from, to, choice: 0 };
                        }
                        return getRangeByType(rangeType);
                    }
                };

                const { from, to, choice } = await getRange(options);
                const productivityByProject = generateReport(data, from, to, options.absolute);
                console.log(formatReport(productivityByProject, choice, from, to, options.absolute));
                break;
            }
            default:
                console.error('Unknown command');
                break;
        }
    } catch (error) {
        console.error(error);
    }
}

main().catch((err) => console.log(err, 'Error while running'))
