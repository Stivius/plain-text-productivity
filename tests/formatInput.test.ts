import { formatInputForFile, formatInputForConsole } from "../src/formatInput";

const DATA_SAMPLE = {
  Key1: "Value1",
  Key2: "Value2",
  Key3: "Value3",
};

describe("formatInput", () => {
  const date = new Date("2022-01-01");

  test("formatInputForFile", () => {
    expect(formatInputForFile(date, DATA_SAMPLE)).toMatchSnapshot();
  });

  test("formatInputForConsole", () => {
    expect(formatInputForConsole(date, DATA_SAMPLE)).toMatchSnapshot();
  });
});
