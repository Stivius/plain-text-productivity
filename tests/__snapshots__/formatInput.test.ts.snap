// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`formatInput formatInputForConsole 1`] = `
"[34m2022-01-01[39m
PROJECT MARK  
[33mKey1[39m    [37mValue1[39m
[33mKey2[39m    [37mValue2[39m
[33mKey3[39m    [37mValue3[39m"
`;

exports[`formatInput formatInputForFile 1`] = `
"2022-01-01
Key1:Value1
Key2:Value2
Key3:Value3

"
`;
